### ISSUE TITLE
Expect-ct - Report Uri form

### ISSUE THREAD

**Issue by AlphaDE** [o10n-x/wordpress-security-header-optimization#7]
I have a problem with configuring Expect-CT

When I try that and add my .php url with the reporting script

https://www.domain.de/expect-ct_reporting.php

I get

 Failed to save settings. The input contains the following errors.

headers.expect-ct The property report-uri is not defined and the definition does not allow additional properties

Any hint?

**Comment by optimalisatie** [comment]
Thank you for reporting the issue!

The configuration of all WPO optimization plugins is 100% stored in JSON and verified using JSON schema's. This makes it reliable with over 10,000 complex optimization settings. There is not much that could break if the JSON schema is valid. You could even use external JSON editors (e.g. Sublime Text) to write optimization config without errors.

### History

Our originally planned **Advanced Optimization** plugin (1 full plugin) contained an advanced JSON config loading and editing innovation that enabled to use millions of JSON optimization documents to fine tune the optimization settings for millions of pages on a website (e.g. a big webshop). This feature would enable an AI service (API) to automatically fine tune the optimization for individual pages and even up to specific device model or individual visitor level. It would be able to take optimization to a new level.

The original plan for the plugin had 3 stages:

1) publication of a high quality Advanced Optimization plugin (Google Lighthouse optimization + beyond)
2) make the JSON optimization config popular (it can save a lot of time, easy to get free help (JSON knowledge) on forums, easy for examples in tutorials etc)
3) use the optimization intelligence data unlocked trough the JSON documents to create an AI automation service (API) that can provide the best possible performance result on per-page, per-device and even individual visitor level.

The AI service would be an external service, independent from the free and open source plugin. The port for the AI would also be open source, allowing other solutions to make use of it's potential for automation.

Stage 1 was completed. Although the plugin is a raw form and has many issues related to usability and documentation, it could be transformed rapidly if it were to be taken further. A lot is possible with configuration based on JSON. It would be possible for example to create a cloud based optimization panel (SaaS) or a semi-automated (AI powered) optimization wizard etc.

Due to issues with the project's funding, the bigger plugin was cut into individual plugins and stripped of many features. Just a part of the JSON configuration innovation was maintained while the configuration is stored in the WordPress options database.

Later, a quick fix solution to achieve some of the originally intended functionality for local JSON document based configuration was added via a **Stealth Optimization Config Proxy**:

https://github.com/o10n-x/wordpress-http2-optimization/releases/tag/0.0.55

In the WPO plugins the JSON config innovation simply provides reliablity by depending completely on JSON schema validation.

In this case the JSON schema contains an older reference `reportUri` that was later renamed to `report-uri` to match the official reference. The JSON schemas are stored centrally and are copied to the plugin core. By mistake, we only edited the plugin core schema and thereby the update wasn't published.

```json
{
    "report-uri": {
        "title": "Report URI",
        "oneOf": [{
            "type": "string",
            "enum": [""]
        }, {
            "type": "string",
            "format": "uri"
        }]
    }
}
```

https://github.com/o10n-x/wordpress-security-header-optimization/blob/master/core/schemas/headers.json

We have updated the plugin with the correct JSON schema.

**Comment by AlphaDE** [comment]
I updated and got v 0.73 for the security plugin (all other's were updated, too), but still get the same error?

**Comment by optimalisatie** [comment]
Thanks a lot for the extended feedback! Your feedback in the past has resulted a great quality improvement for the WPO plugins.

At our test installation, specifically for the CSS optimization plugin, Github updater downloads `0.0.99` while advertising `0.0.100`. Since `C` is the first letter, the plugin core shipped with the CSS optimization plugin is likely shared by all other plugins. Thereby, the security optimization plugin does not have access to the latest JSON schema files.

![image](https://user-images.githubusercontent.com/8843669/56208446-77339f00-6051-11e9-88b1-78df11a0e08e.png)

![image](https://user-images.githubusercontent.com/8843669/56208516-9b8f7b80-6051-11e9-8766-8013a4f4dc1f.png)

Regrettably, we already tried to upgrade to 0.0.101 and while Github updater does recognize the new version `0.0.101`, it starts a download for `0.0.99`. Perhaps it is a bug in Github Updater that does not support a number higher than 99.

It would require an investigation. To speed things up, it would be an option to ask for support at Github Updater's forum. Regrettably, we currently do not have time available.

When newly installing the plugin via Github updater it does work. However, I noticed that it uses the `master` tag instead of `0.0.100` to install the plugin. 

![image](https://user-images.githubusercontent.com/8843669/56209644-56207d80-6054-11e9-8478-6e6b4c3f57b7.png)

Perhaps the bug is known at Github Updater.

https://github.com/afragen/github-updater/issues

**Comment by AlphaDE** [comment]
Jan, I adressed the Github updater problem and opened an issue there.

Nevertheless, the Security plugin is current 0.73 and I still get the error regarding the wrong reporting URI for Expect-CT

**Comment by AlphaDE** [comment]
Ok, the css plugin was blocking this, I manually installed it and got it fixed.

**Comment by AlphaDE** [comment]
One more thing regarding Expect-CT: For separation, semicolon are used, the specification says there need to be commas.

The policy provided by the server is invalid and can't be parsed. Please refer to the Expect-CT specification for more information. If you're configured your policy based on an older version of the specification, please note that newer versions use commas to separate directives. URLs should be enclosed in double quotes.

**Comment by optimalisatie** [comment]
The issue has been fixed in `v0.0.74`.

https://github.com/o10n-x/wordpress-security-header-optimization/releases

**Comment by AlphaDE** [comment]
Great, thanks.

**Comment by AlphaDE** [comment]
Github updater author replied:

So here’s the problem. The developer tagged a release as 0.99 and unless they delete it, a higher release will need to be 0.991 or greater. 0.0.100 is lower than 0.99 and that’s why 0.99 is being served.

**Comment by strayangelfilms** [comment]
New version should likely be 0.1.00 instead of 0.0.100.

**Comment by AlphaDE** [comment]
Yes, or if you plan for subversion numbers >100, you need to start with three digits from the very beginning... .001 instead of .01 because it looks like a string comparison is used where 0.99 >0.100 instead of a number conversion where 0.99 < 0.100 (when the digits are separated, hard to code).

**Comment by afragen** [comment]
Hi, developer if GitHub Updater here. 

The problem seems to be a tagged release of 0.99 was created in addition to a tagged release of 0.0.99. To make everything work as expected again you will need to delete the tagged 0.99

https://github.com/o10n-x/wordpress-css-optimization/releases/tag/0.99

**Comment by optimalisatie** [comment]
Hi @afragen,

You were correct. We made a mistake when creating a Github release. 

Our sincere apologies for filing a false bug report. And thanks a lot for the great support!

The incorrect tag has been deleted. The update is now working.

![image](https://user-images.githubusercontent.com/8843669/56265384-a990db00-60e9-11e9-881c-48380f65afd1.png)

Thanks a lot for providing the Github Updater! From the feedback that we received it has become evident that many people are very happy with it and that it is working well.

@strayangelfilms thank you for your suggestion. I've included you since the message contained information regarding the history of PSP (and since your account is part of the project's team). You are free to share your perspective if you want to provide clarity in regards to PSP and the project. Despite our agreement, you are still the official CEO of the project.
